[Build] Unable to use CoreML: This binary was not compiled with CoreML support.

### Describe the issue

Hello, I am using the dependency - 

```
    <dependency>
      <groupId>com.microsoft.onnxruntime</groupId>
      <artifactId>onnxruntime</artifactId>
      <version>1.15.0</version>
    </dependency>

```
and with the following code- 
``` java
  var env = OrtEnvironment.getEnvironment();
  var sessionOptions = new OrtSession.SessionOptions();
  sessionOptions.addCoreML();
  var session = env.createSession("model.onnx", sessionOptions);
```

I get error 

```Error code - ORT_INVALID_ARGUMENT - message: This binary was not compiled with CoreML support.```

So I tried building the package myself like this- 
`./build.sh --update --build --build_java --config Release --parallel`

but that gives an error 

```
CMake Error at /opt/homebrew/Cellar/cmake/3.26.4/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find JNI (missing: JAVA_INCLUDE_PATH JAVA_INCLUDE_PATH2 AWT)
Call Stack (most recent call first):
  /opt/homebrew/Cellar/cmake/3.26.4/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  /opt/homebrew/Cellar/cmake/3.26.4/share/cmake/Modules/FindJNI.cmake:569 (find_package_handle_standard_args)
  onnxruntime_java.cmake:11 (find_package)
  CMakeLists.txt:1631 (include)
  ```

`java -version` output- 

```
openjdk version "17.0.7" 2023-04-18
OpenJDK Runtime Environment Homebrew (build 17.0.7+0)
OpenJDK 64-Bit Server VM Homebrew (build 17.0.7+0, mixed mode, sharing)
```

Any suggestion on how to go about fixing it? I would rather use pre-built binary than spend time resolving build issues 



### Urgency

_No response_

### Target platform

macOS

### Build script

`./build.sh --update --build --build_java --config Release --parallel`

### Error / output

`  Could NOT find JNI (missing: JAVA_INCLUDE_PATH JAVA_INCLUDE_PATH2 AWT)
Call Stack (most recent call first):
  /opt/homebrew/Cellar/cmake/3.26.4/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  /opt/homebrew/Cellar/cmake/3.26.4/share/cmake/Modules/FindJNI.cmake:569 (find_package_handle_standard_args)
  onnxruntime_java.cmake:11 (find_package)
  CMakeLists.txt:1631 (include)`

### Visual Studio Version

_No response_

### GCC / Compiler Version

arm64-apple-darwin22.4.0

Thanks so much for the detailed response, seems like `JAVA_HOME` and `cmake` are acting naughty on my machine. 
if I do `ls $JAVA_HOME/include` I see 
```
classfile_constants.h jawt.h                jawt_md.h             jdwpTransport.h       jni.h                 jni_md.h              jvmti.h               jvmticmlr.h
```

but when I run `./build.sh --update --build --build_java --config Release --parallel --use_coreml` I still get this error- 

```
CMake Error at /Applications/CMake.app/Contents/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find JNI (missing: JAVA_INCLUDE_PATH JAVA_INCLUDE_PATH2 AWT)
Call Stack (most recent call first):
  /Applications/CMake.app/Contents/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  /Applications/CMake.app/Contents/share/cmake-3.26/Modules/FindJNI.cmake:569 (find_package_handle_standard_args)
  onnxruntime_java.cmake:16 (find_package)
  CMakeLists.txt:1631 (include)
```
so it seems like `JAVA_HOME` is not being honored by `cmake` or it is somehow not able to read it. There is some reference to the problem [here](https://stackoverflow.com/questions/53876767/cmake-not-finding-jni-when-java-home-is-set-correctly). 

If you don't mind, @Craigacp , can you please share what is the output of  ` ls $JAVA_HOME` on your machine? 


Also, note that I was suspecting something wrong with `cmake` installed through `brew` so I installed the binary from cmake official site, that's why you see `/Applications/CMake.app/Contents/share/cmake` in the error log I posted. 


On a different note, are there plans of shipping CoreML JNI binary through Maven packages? Given there are already binaries available in the jar file, I think it will be a good idea - a lot of people will be running into build issues otherwise.

Gotcha. I tried cleaning up cmake cache, but now I am getting the error- 

```
/Users/<redacted>/onnxruntime/build/MacOS/Release/_deps/flatbuffers-src/src/idl_gen_rust.cpp:409:12: error: variable 'i' set but not used [-Werror,-Wunused-but-set-variable]
    size_t i = 0;
    ```
    
The command I am running is `./build.sh  --build_java --config Release --parallel --use_coreml` ; it _feels_ like the code is auto-generated so I am hesitant to hack my way around it. 

Any suggestions on how to go about fixing it?

> Gotcha. I tried cleaning up cmake cache, but now I am getting the error-
> 
> ```
> /Users/<redacted>/onnxruntime/build/MacOS/Release/_deps/flatbuffers-src/src/idl_gen_rust.cpp:409:12: error: variable 'i' set but not used [-Werror,-Wunused-but-set-variable]
>     size_t i = 0;
> ```
> 
> The command I am running is `./build.sh --build_java --config Release --parallel --use_coreml` ; it _feels_ like the code is auto-generated so I am hesitant to hack my way around it.
> 
> Any suggestions on how to go about fixing it?

ORT doesn't build on Xcode 14.3 now without some modifications. Here's a workaround for that particular issue:

https://github.com/microsoft/onnxruntime/pull/15878/files#diff-cf577ff910eb6272a6a671d1c3eeb1b3b5a6d24357332306e37f8ea533b8cf2d

#15878 should get things building again.

--- a/java/build.gradle
+++ b/java/build.gradle
@@ -181,7 +181,7 @@ test {
 	if (cmakeBuildDir != null) {
 		workingDir cmakeBuildDir
 	}
-	systemProperties System.getProperties().subMap(['USE_CUDA', 'USE_ROCM', 'USE_TENSORRT', 'USE_DNNL', 'USE_OPENVINO', 'JAVA_FULL_TEST', 'ENABLE_TRAINING_APIS'])
+	systemProperties System.getProperties().subMap(['USE_CUDA', 'USE_ROCM', 'USE_TENSORRT', 'USE_DNNL', 'USE_OPENVINO', 'USE_COREML', 'JAVA_FULL_TEST', 'ENABLE_TRAINING_APIS'])
 	testLogging {
 		events "passed", "skipped", "failed"
 		showStandardStreams = true

--- a/java/src/main/native/ai_onnxruntime_OrtSession_SessionOptions.c
+++ b/java/src/main/native/ai_onnxruntime_OrtSession_SessionOptions.c
@@ -600,7 +600,7 @@ JNIEXPORT void JNICALL Java_ai_onnxruntime_OrtSession_00024SessionOptions_addArm
 JNIEXPORT void JNICALL Java_ai_onnxruntime_OrtSession_00024SessionOptions_addCoreML
   (JNIEnv * jniEnv, jobject jobj, jlong apiHandle, jlong handle, jint coreMLFlags) {
     (void)jobj;
-  #ifdef USE_CORE_ML
+  #ifdef USE_COREML
     checkOrtStatus(jniEnv,(const OrtApi*)apiHandle,OrtSessionOptionsAppendExecutionProvider_CoreML((OrtSessionOptions*) handle, (uint32_t) coreMLFlags));
   #else
     (void)apiHandle;(void)handle;(void)coreMLFlags; // Parameters used when CoreML is defined.

--- a/java/src/test/java/ai/onnxruntime/InferenceTest.java
+++ b/java/src/test/java/ai/onnxruntime/InferenceTest.java
@@ -632,6 +632,12 @@ public class InferenceTest {
     runProvider(OrtProvider.XNNPACK);
   }
 
+  @Test
+  @EnabledIfSystemProperty(named = "USE_COREML", matches = "1")
+  public void testCoreML() throws OrtException {
+    runProvider(OrtProvider.CORE_ML);
+  }
+
   private void runProvider(OrtProvider provider) throws OrtException {
     EnumSet<OrtProvider> providers = OrtEnvironment.getAvailableProviders();
     assertTrue(providers.size() > 1);

